components/sup/src/manager/file_watcher.rs

```diff
@@ -0,0 +1,1463 @@
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use std::env;
+use std::collections::{HashMap, HashSet, VecDeque};
+use std::collections::hash_map::Entry;
+use std::ffi::OsString;
+use std::mem::swap;
+use std::path::{Component, Path, PathBuf};
+use std::sync::mpsc::{Receiver, channel};
+use std::time::Duration;
+
+use error::{Error, Result};
+use notify;
+use notify::{DebouncedEvent, RecommendedWatcher, RecursiveMode, Watcher};
+
+const WATCHER_DELAY_MS: u64 = 2_000;
+static LOGKEY: &'static str = "FW";
+
+/// A set of callbacks for the watched file events.
+pub trait Callbacks {
+    /// A function that gets called when the watched file shows up.
+    ///
+    /// `real_path` will contain a path to the watched file, which may
+    /// be different from the one that was passed to `FileWatcher`,
+    /// because of symlinks.
+    fn file_appeared(&mut self, real_path: &Path);
+    /// A function that gets called when the watched file is written to.
+    ///
+    /// Note that this is called only when the real file is
+    /// modified. In case when some symlink in the watched paths is
+    /// atomically changed to point to something else,
+    /// `file_disappeared` followed by `file_appeared` will be
+    /// actually called.
+    ///
+    /// `real_path` will contain a path to the watched file, which may
+    /// be different from the one that was passed to `FileWatcher`,
+    /// because of symlinks.
+    fn file_modified(&mut self, real_path: &Path);
+    /// A function that gets called when the watched file goes away.
+    ///
+    /// `real_path` will contain a path to the watched file, which may
+    /// be different from the one that was passed to `FileWatcher`,
+    /// because of symlinks.
+    fn file_disappeared(&mut self, real_path: &Path);
+}
+
+// Essentially a pair of dirname and basename.
+#[derive(Clone, Debug, Default)]
+struct DirFileName {
+    directory: PathBuf,
+    file_name: OsString,
+}
+
+impl DirFileName {
+    // split_path separates the dirname from the basename.
+    fn split_path(path: &PathBuf) -> Option<Self> {
+        let parent = match path.parent() {
+            None => return None,
+            Some(p) => p,
+        };
+        let file_name = match path.file_name() {
+            None => return None,
+            Some(f) => f,
+        };
+        Some(Self {
+            directory: parent.to_owned(),
+            file_name: file_name.to_owned(),
+        })
+    }
+
+    fn as_path(&self) -> PathBuf {
+        self.directory.join(&self.file_name)
+    }
+}
+
+// TODO: handle mount events, we could use libc crate to get the
```

Comment:
Think a `TODO(krnowak)` would be useful here, so people know who to ping in case they have questions.